bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitCharacteristicFactory.java

```diff
@@ -195,315 +200,330 @@ private static void setValueFromEnum(GenericItem item, CharacteristicEnum value,
     }
 
-    private static LockPhysicalControlsCharacteristic createLockPhysicalControlsCharacteristic(final GenericItem item,
-            HomekitAccessoryUpdater updater) {
+    private static LockPhysicalControlsCharacteristic createLockPhysicalControlsCharacteristic(
+            final HomekitTaggedItem item, HomekitAccessoryUpdater updater) {
         return new LockPhysicalControlsCharacteristic(
                 () -> getEnumFromItem(item, LockPhysicalControlsEnum.CONTROL_LOCK_DISABLED,
                         LockPhysicalControlsEnum.CONTROL_LOCK_ENABLED, LockPhysicalControlsEnum.CONTROL_LOCK_DISABLED),
                 (value) -> setValueFromEnum(item, value, LockPhysicalControlsEnum.CONTROL_LOCK_DISABLED,
                         LockPhysicalControlsEnum.CONTROL_LOCK_ENABLED),
-                (callback) -> updater.subscribe(item, LOCK_CONTROL.getTag(), callback),
-                () -> updater.unsubscribe(item, LOCK_CONTROL.getTag()));
+                getSubscriber(item, LOCK_CONTROL, updater), getUnsubscriber(item, LOCK_CONTROL, updater));
     }
 
-    private static RotationSpeedCharacteristic createRotationSpeedCharacteristic(final GenericItem item,
+    private static RotationSpeedCharacteristic createRotationSpeedCharacteristic(final HomekitTaggedItem item,
             HomekitAccessoryUpdater updater) {
         return new RotationSpeedCharacteristic(getIntSupplier(item), setIntConsumer(item),
-                (callback) -> updater.subscribe(item, ROTATION_SPEED.getTag(), callback),
-                () -> updater.unsubscribe(item, ROTATION_SPEED.getTag()));
+                getSubscriber(item, ROTATION_SPEED, updater), getUnsubscriber(item, ROTATION_SPEED, updater));
     }
 
-    private static SetDurationCharacteristic createDurationCharacteristic(final GenericItem item,
+    private static SetDurationCharacteristic createDurationCharacteristic(final HomekitTaggedItem item,
             HomekitAccessoryUpdater updater) {
-        return new SetDurationCharacteristic(getIntSupplier(item), setIntConsumer(item),
-                (callback) -> updater.subscribe(item, DURATION.getTag(), callback),
-                () -> updater.unsubscribe(item, DURATION.getTag()));
+        return new SetDurationCharacteristic(() -> {
+            int value = 0;
+            final State state = item.getItem().getState();
+            if (state instanceof PercentType) {
+                value = state.as(PercentType.class).intValue();
+            } else if (state instanceof DecimalType) {
+                value = state.as(DecimalType.class).intValue();
+            } else if (state instanceof UnDefType) {
+                logger.debug("Item state {} is UNDEF {}.", state, item.getName());
+            } else {
+                logger.warn(
+                        "Item state {} is not supported for {}. Only PercentType and DecimalType (0/100) are supported.",
+                        state, item.getName());
+            }
```

Comment:
I don't like that all this code is duplicated from `getIntSupplier`. Can you refactor it?

Reply:
yes, i had to add few lines to getIntSupplier. but indeed, it can be solved much better than copy-paste. 
i will split getIntSupplier in 2 parts - getInt and getSupplier so that i can use getInt here and add  required lines.